packages/api/src/external/hie/register-facility.ts.ts

```diff
@@ -0,0 +1,110 @@
+import {
+  AddressStrict,
+  AddressWithCoordinates,
+  removeCoordinates,
+} from "@metriport/core/domain/location-address";
+import {
+  FacilityRegister,
+  Facility,
+  FacilityCreate,
+  isOboFacility,
+} from "../../domain/medical/facility";
+import { buildCqOrgName } from "../../external/carequality/shared";
+import { buildCwOrgName } from "../../external/commonwell/shared";
+import { getCxOrganizationNameOidAndType } from "../../command/medical/organization/get-organization";
+import { getAddressWithCoordinates } from "../../domain/medical/address";
+import { getCqOboData } from "../../external/carequality/get-obo-data";
+import { createOrUpdateFacility } from "../../command/medical/facility/create-or-update-facility";
+import { createOrUpdateInCq, createOrUpdateInCw } from "./shared";
+
+/**
+ * Registers a new facility within HIEs.
+ *
+ * @param cxId
+ * @param facility
+ * @returns The updated facility.
+ */
+export async function registerFacilityWithinHIEs(
+  cxId: string,
+  facility: FacilityRegister
+): Promise<Facility> {
+  const isObo = isOboFacility(facility.type);
+
+  const [cxOrg, address, cqOboData] = await Promise.all([
+    getCxOrganizationNameOidAndType(cxId),
+    getAddressWithCoordinates(getAddressFromInput(facility), cxId),
+    getCqOboData(facility.cqOboActive, facility.cqOboOid),
+  ]);
+
+  const facilityDetails = createFacilityDetails(cxId, facility, address);
+
+  let oboFacilityDetails = facilityDetails;
+
+  if (isObo) {
+    oboFacilityDetails = {
+      ...oboFacilityDetails,
+      cqOboActive: facility.cqOboActive,
+      cqOboOid: facility.cqOboOid,
+      cwOboActive: facility.cwOboActive,
+      cwOboOid: facility.cwOboOid,
+    };
+  }
+
+  const cmdFacility = await createOrUpdateFacility(
+    cxId,
+    facility.id,
+    facility.data.npi,
+    oboFacilityDetails
+  );
+
+  // CAREQUALITY
+  const cqOboEnabled = isObo && cqOboData.enabled;
+  if (cqOboEnabled || !isObo) {
+    const cqFacilityName = cqOboEnabled ? cqOboData.cqFacilityName : cmdFacility.data.name;
+    const cqOboOid = cqOboEnabled ? cqOboData.cqOboOid : undefined;
+    const cqOrgName = buildCqOrgName(cxOrg.name, cqFacilityName, cqOboOid);
+    await createOrUpdateInCq(cmdFacility, cxOrg.oid, cqOrgName, address);
```

Comment:
The way we're dealing w/ address on this function is prone to future errors as we update the codebase.

We get the address from `getAddressWithCoordinates` and pass it to the function that creates the decides the facility details (`createFacilityDetails`);

Then we use that address again here. At first glance, that seems ok.

But let's say we decide in the future to update `createFacilityDetails` to do something w/ address and it ends updating it (e.g., it returns the org's address instead of the facility's).

Once we pass the responsibility to decide the facility's data/details to `createFacilityDetails`, we should treat its response as the source of truth - the address from the input and from `getAddressWithCoordinates` might be outdated.

I looked at the functions/returns, we could make the following changes that should be simple:
1. `removeCoordinates()` returns `{ address: AddressStrict; coordinates: Coordinates; }`
1. then `createFacilityDetails()` can return `{ facilityDetails: FacilityCreate; coordinates: Coordinates }` (the now source of truth for the facility data)
1. `createOrUpdateInCq` gets the coordinates from (2)
   - could even simplify its 4th param to be `coordinates: Coordinates` 

Not a blocker, but I suggest the update to make the code more robust. Could ticket and link it on the code as TODO if you prefer.